File: clients/go/consensus/compactsize.go
package consensus

func readCompactSize(b []byte, off *int) (uint64, int, error) {
	start := *off
	tag, err := readU8(b, off)
	if err != nil {
		return 0, 0, err
	}
	v, err := readCompactSizeTagged(tag, b, off)
	if err != nil {
		return 0, 0, err
	}
	return v, *off - start, nil
}

func readCompactSizeTagged(tag byte, b []byte, off *int) (uint64, error) {
	if tag < 0xfd {
		return uint64(tag), nil
	}
	switch tag {
	case 0xfd:
		return readCompactSize16(b, off)
	case 0xfe:
		return readCompactSize32(b, off)
	case 0xff:
		return readCompactSize64(b, off)
	default:
		return 0, txerr(TX_ERR_PARSE, "invalid CompactSize tag")
	}
}

func readCompactSize16(b []byte, off *int) (uint64, error) {
	v, err := readU16le(b, off)
	if err != nil {
		return 0, err
	}
	if v < 0xfd {
		return 0, txerr(TX_ERR_PARSE, "non-minimal CompactSize (0xfd)")
	}
	return uint64(v), nil
}

func readCompactSize32(b []byte, off *int) (uint64, error) {
	v, err := readU32le(b, off)
	if err != nil {
		return 0, err
	}
	if v <= 0xffff {
		return 0, txerr(TX_ERR_PARSE, "non-minimal CompactSize (0xfe)")
	}
	return uint64(v), nil
}

func readCompactSize64(b []byte, off *int) (uint64, error) {
	v, err := readU64le(b, off)
	if err != nil {
		return 0, err
	}
	if v <= 0xffff_ffff {
		return 0, txerr(TX_ERR_PARSE, "non-minimal CompactSize (0xff)")
	}
	return v, nil
}

File: clients/rust/crates/rubin-consensus/src/compactsize.rs
use crate::error::{ErrorCode, TxError};
use crate::wire_read::Reader;

pub fn read_compact_size(r: &mut Reader<'_>) -> Result<(u64, usize), TxError> {
    let start = r.offset();
    let tag = r.read_u8()?;

    let (v, minimal_ok) = match tag {
        0x00..=0xfc => (tag as u64, true),
        0xfd => {
            let v = r.read_u16_le()? as u64;
            (v, v >= 0xfd)
        }
        0xfe => {
            let v = r.read_u32_le()? as u64;
            (v, v > 0xffff)
        }
        0xff => {
            let v = r.read_u64_le()?;
            (v, v > 0xffff_ffff)
        }
    };

    if !minimal_ok {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "non-minimal CompactSize",
        ));
    }

    Ok((v, r.offset() - start))
}

pub fn read_compact_size_bytes(b: &[u8]) -> Result<(u64, usize), TxError> {
    let mut r = Reader::new(b);
    read_compact_size(&mut r)
}

pub fn encode_compact_size(n: u64, out: &mut Vec<u8>) {
    match n {
        0x00..=0xfc => out.push(n as u8),
        0xfd..=0xffff => {
            out.push(0xfd);
            out.extend_from_slice(&(n as u16).to_le_bytes());
        }
        0x1_0000..=0xffff_ffff => {
            out.push(0xfe);
            out.extend_from_slice(&(n as u32).to_le_bytes());
        }
        _ => {
            out.push(0xff);
            out.extend_from_slice(&n.to_le_bytes());
        }
    }
}

// ---------------------------------------------------------------------------
// Kani bounded model checking proofs
// ---------------------------------------------------------------------------
#[cfg(kani)]
mod verification {
    use super::*;

    /// For any u64 value, encode → decode must roundtrip to the same value
    /// and consume exactly the number of bytes produced.
    #[kani::proof]
    fn verify_compact_size_roundtrip() {
        let n: u64 = kani::any();
        let mut buf = Vec::new();
        encode_compact_size(n, &mut buf);

        let (decoded, consumed) = read_compact_size_bytes(&buf).unwrap();
        assert_eq!(decoded, n);
        assert_eq!(consumed, buf.len());
    }

    /// encode_compact_size always produces minimal encoding (no wasted bytes).
    #[kani::proof]
    fn verify_compact_size_encoding_is_minimal() {
        let n: u64 = kani::any();
        let mut buf = Vec::new();
        encode_compact_size(n, &mut buf);

        let expected_len = match n {
            0x00..=0xfc => 1,
            0xfd..=0xffff => 3,
            0x1_0000..=0xffff_ffff => 5,
            _ => 9,
        };
        assert_eq!(buf.len(), expected_len);
    }

    /// Decoding any 9-byte buffer never panics (returns Ok or Err, no UB).
    #[kani::proof]
    fn verify_compact_size_decode_no_panic() {
        let buf: [u8; 9] = kani::any();
        let _ = read_compact_size_bytes(&buf);
    }
}
